### PR TITLE
Compute the actual endKey for the scan.

### PR DESCRIPTION
**Pull request by petermattis**
This avoids pathological behavior with the timestamp cache when always
scanning to a fixed end key, more closely resolves scan usage in the
cockroach codebase and expected usage from real world clients.

Change the benchmark to use the RPC client. The old and new numbers
below are both using the RPC client.

benchmark                               old ns/op     new ns/op     delta
BenchmarkClientScan1Version1Row         328182        301983        -7.98%

benchmark                               old MB/s     new MB/s     speedup
BenchmarkClientScan1Version1Row         3.12         3.39         1.09x

### CODE REVIEW

File: client/client_test.go [hunk 1]
```diff
@@ -806,12 +805,14 @@ func runClientScan(numRows, numVersions int, b *testing.B) {
...
-		keyBuf := append(make([]byte, 0, 64), []byte("key-")...)
-		endKey := []byte("key.")
+		startKeyBuf := append(make([]byte, 0, 64), []byte("key-")...)
+		endKeyBuf := append(make([]byte, 0, 64), []byte("key-")...)
+		// endKey := []byte("key.")
 		for pb.Next() {
 			// Choose a random key to start scan.
 			keyIdx := rand.Int31n(int32(numKeys - numRows))
-			startKey := proto.Key(encoding.EncodeUvarint(keyBuf[0:4], uint64(keyIdx)))
+			startKey := proto.Key(encoding.EncodeUvarint(startKeyBuf[0:4], uint64(keyIdx)))
```
Comment (mrtracy) [inline client/client_test.go]: leave out the 0 if buffer slices start from the beginning: `startKeyBuf[:4]`

I think one of gofmt's modes makes this change automatically.

Comment (mrtracy) [inline client/client_test.go]: Actually, I don't quite get why you're taking a slice at all.

**Comment by spencerkimball** [comment]
LGTM

File: client/client_test.go (hunk 1 above)
Comment (petermattis) [inline client/client_test.go]: You're right that there is no need to slice the bufs. I sometimes get confused that operations which take a slice do not modify the slice but return a new slice.

**Pull request merged**